### PR TITLE
Created steppable Optimizer and steppable gradient descent

### PR DESCRIPTION
**Pull request by MarcDrudis**
This pull request has the objective of start developping steppable optimizers before making a pull request into Qiskit/main.

### CODE REVIEW

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        self._state = None
+        self.maxiter = maxiter
+    
+    @abstractmethod
+    def ask(self)-> AskObject:
+        """
+        Abstract method ask. This method is the part of the interface of the optimizer that asks the 
+        user/quantum_circuit how and where the function to optimize needs to be evaluated. It is the
+        first method inside of a "step" in the optimization process.
+        """
```
Comment (eggerdj) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]:
```suggestion
        """Ask the optimizer for a set of points to evaluate.
        
        This method asks the optimizer which are the next points to evaluate. 
        These points can, e.g., correspond to function values and/or its derivative. 
        It may also correspond to variables that let the user infer which points to evaluate.
        It is the first method inside of a "step" in the optimization process.
        """
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        raise NotImplementedError
+    @abstractmethod
+    def tell(self,tell_object:TellObject)->None:
```
Comment (eggerdj) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: Needs the AskObject

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
+from abc import abstractmethod,ABC
+from dataclasses import dataclass
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List
+
+from .optimizer import Optimizer, POINT, OptimizerResult
+
+@dataclass
+class AskObject(ABC):
+    """Base class for return type of method SteppableOptimizer.ask()"""
```
Comment (eggerdj) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: This can be different from the `x` in the `OptimizerState`. Worth mentioning.

File: qiskit/algorithms/optimizers/steppable_gradient_descent.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        objective_function,
+        gradient_function=None,
+        variable_bounds=None,
+        initial_point=None,
+    ):
+        super().optimize(
+            num_vars, objective_function, gradient_function, variable_bounds, initial_point
+        )
+        result = self.minimize(objective_function, initial_point, gradient_function)
+        return result.x, result.fun, result.nfev
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_gradient_descent.py:177]: This can be removed 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            ask_object: Contains the information on how to do the evaluation.
+        Return:
+            tell_object: Contains all relevant information about the evaluation of the objective function.
+        """
+        raise NotImplementedError
+
+    def user_evaluate(
+        self, *args, **kwargs
+    ) -> TellObject:  # Maybe this function should be abstract. For some optimizers (The return type with a custom tell object the function needs to be overwriten)
+        """
```
Comment (eggerdj) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:176]: The doc needs to explain why this is needed.

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    @abstractmethod
+    def create_result(self) -> OptimizerResult:
+        """
+        Returns the result of the optimization.
+        All the information needed to create the result should be stored in the optimizer state.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def initialize(self, *args, **kwargs) -> None:
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: I think this should have the same signature as `minimize`, so that that the following two usages are the same:
```python
result = opt.minimize(f, x0)
```
and 
```python
opt.initialize(f, x0)
while not opt.stop_criterion():
     opt.step()
```
All the hyperparameters are currently stored in the `Optimizer.__init__` and to not restrict hyperparameter setting to `initialize` we should keep it this way 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        if self.callback is not None:
+            self.callback(state=self._state)
+
+    def minimize(
+        self,
+        fun: Callable[[POINT], float],
+        x0: POINT,
+        jac: Optional[Callable[[POINT], POINT]] = None,
+        bounds: Optional[List[Tuple[float, float]]] = None,
+        **kwargs,
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]:
```suggestion
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        """
+        For well behaved functions the user can call this method to minimize a function. If the user wants more control
+        on how to evaluate the function a custom loop can be created using ask() and tell() and evaluating the function
+        manually.
+        """
+        self.initialize(x0=x0, fun=fun, jac=jac)
+        for _ in range(self.maxiter):
+            self.step()
+            if self.stop_condition():
+                break
```
Comment (eggerdj) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: Alternatively
```suggestion
        while self.continue():
            self.step()
```

File: qiskit/algorithms/optimizers/optimizer.py [hunk 9]
```diff
@@ -92,7 +92,7 @@ def njev(self, njev: Optional[int]) -> None:
     @property
     def nit(self) -> Optional[int]:
         """The total number of iterations."""
-        return self._njev
+        return self._nit
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/optimizer.py:95]: Good catch! This is also included in Qiskit/qiskit-terra#8123

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    the function next, but depending on the optimizer it can also be about whether we should evaluate the function itself or its gradient.
+
+    Once the function has been evaluated, the user calls the method tell() to tell the optimizer what has been the result of the function evaluation.
+    The optimizer then updates its state accordingly and the user can decide whether to stop the optimization process or to repeat a step.
+
+    This interface is more customizable, and allows the user to have full control over the evaluation of the function.
+
+    For example:
+    .. code-block::python
+
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:77]: Could you include all the imports needed for this to run so that users can just copy+paste this snippet? 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+from .optimizer import Optimizer, POINT, OptimizerResult
+import numpy as np
+
+CALLBACK = Callable[[int, np.ndarray, float, float], None]
+
+
+@dataclass
+class AskObject(ABC):
+    """Base class for return type of method SteppableOptimizer.ask()"""
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: Could you add class field documentations that state what `x_fun` and `x_jac` are?

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+    x_fun: Union[POINT, List[POINT]]
+    x_jac: Union[POINT, List[POINT]]
+
+
+@dataclass
+class TellObject(ABC):
+    """Base class for argument type of method SteppableOptimizer.tell()"""
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: Also here some docs explaining the class fields would be helpful I think 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+    nfev: Optional[int]
+    njev: Optional[int]
+    nit: Optional[int]
+
+
+class SteppableOptimizer(Optimizer):
+    """
+    Base class for a steppable optimizer.
+
+    This family of optimizers will be using the ask and tell interface.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: Could you reference the link from below already here? 
```suggestion
    This family of optimizers will be using the 
    `ask and tell interface <https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html>`_.
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+        optimizer = SteppableGradientDescent(maxiter=20)
+        optimizer.initialize(x0=initial_point, fun=objective, jac=grad)
+
+        for _ in range(20):
+            ask_object = optimizer.ask()
+            evaluated_gradient = None
+
+            while evaluated_gradient is None:
+                evaluated_gradient = grad(ask_object.x_center)
+                optimizer._state.njev += 1
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:102]: How is this different from just the following?
```suggestion
            evaluated_gradient = grad(ask_object.x_jac)
            optimizer._state.njev += 1
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+
+    """
+
+    def __init__(
+        self,
+        callback: Optional[CALLBACK] = None,
+        maxiter: int = None,
+
+    ):
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: We usually put the `maxiter` as first argument since it's more often specified than the callback 🙂 
```suggestion
        maxiter: Optional[int] = None,
        callback: Optional[CALLBACK] = None,
    ) -> None:
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        """
+        Args:
+            maxiter: Number of steps in the optimization process before ending the loop.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:128]: Could you add the docs for the callback here?

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+        self.callback = callback
+        self.maxiter = maxiter  # Remove maxiter
+
+    def ask(self) -> AskObject:
+        """
+        Returns the information on how to evaluate the objective function.
+        Canonical optimization workflow using ask() and tell() can be seen in SteppableOptimizer.step().
+        Returns:
+            AskOjbect: Since the way to evaluate the function can vary much with different optimization algorithms, the object will be
+        a custom dataclass for each optimizer.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]: Since we already have the type documented in the signature, we don't need it here anymore I think 🙂 
```suggestion

        Returns:
            Since the way to evaluate the function can vary much with different 
            optimization algorithms, the object will be a custom dataclass for each optimizer.
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+    @abstractmethod
+    def evaluate(self, ask_object: AskObject) -> TellObject:
+        """
+        Evaluates the function according to the instructions contained in ask_object.
+        If the user decides to use step() instead of ask() and tell() this function will contain the logic on how to evaluate
+        the function.
+        Args:
+            ask_object: Contains the information on how to do the evaluation.
+        Return:
+            tell_object: Contains all relevant information about the evaluation of the objective function.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]:
```suggestion
            Contains all relevant information about the evaluation of the objective function.
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py (hunk 5 above)
Comment (Cryoris) [reply qiskit/algorithms/optimizers/steppable_optimizer.py:176]: From our discussion I think this method can be removed entirely 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+        """
+        Constructs TellObject.
+        Used when the user manually evaluates the function.
+        """
+        return TellObject(*args, **kwargs)
+
+    def step(self) -> None:
+        """
+        Performs one step in the optimization process.
+        This mehtod composes ask(), evaluate(), and tell() to make a step in the optimization process.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py]:
```suggestion
        This method composes ask(), evaluate(), and tell() to make a step in the optimization process.
```

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+    def minimize(
+        self,
+        fun: Callable[[POINT], float],
+        x0: POINT,
+        jac: Optional[Callable[[POINT], POINT]] = None,
+        bounds: Optional[List[Tuple[float, float]]] = None,
+        **kwargs,
+    ) -> OptimizerResult:
+        """
+        For well behaved functions the user can call this method to minimize a function. If the user wants more control
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:221]: Could you add the argument & return value docs here? 🙂

File: qiskit/algorithms/optimizers/steppable_optimizer.py [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+            if self.stop_condition():
+                break
+
+        return self.create_result()
+
+    @abstractmethod
+    def create_result(self) -> OptimizerResult:
+        """
+        Returns the result of the optimization.
+        All the information needed to create the result should be stored in the optimizer state.
```
Comment (Cryoris) [inline qiskit/algorithms/optimizers/steppable_optimizer.py:240]: This misses return value docs 🙂

**Review by Cryoris — COMMENTED** [review]
I left some comments below, but other than that the base class LGTM 👍🏻

**Comment by MarcDrudis** [comment]
Callback has a different signature depending on the optimization algorithm. I don't know how I will handle, but I wanted to leave a message here to point it out.
For example, in gradient descent `CALLBACK = Callable[[int, np.ndarray, float, float], None]` but in spsa `Callable[[int, np.ndarray, float, float, bool], None]`.

**Comment by MarcDrudis** [comment]
I just realized that minimize() and other methods are already defined in the parent class. For those, should I copy-paste the documetation? Modify it if necessary? Or just leave them without documentation and they will inherit the documentation from the abstract method?

**Comment by MarcDrudis** [comment]
In one of the meetings we talked about where should store the function and the gradient function. Let's talk about it next time as well.
